modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java

```diff
@@ -210,27 +239,109 @@ private synchronized void releaseCachedClients() {
+        AWSCredentialsProvider getCredentialsProvider() {
+            return credentialsProvider;
         }
 
         @Override
         public AWSCredentials getCredentials() {
-            return SocketAccess.doPrivileged(credentials::getCredentials);
+            return SocketAccess.doPrivileged(credentialsProvider::getCredentials);
         }
 
         @Override
         public void refresh() {
-            SocketAccess.doPrivilegedVoid(credentials::refresh);
+            SocketAccess.doPrivilegedVoid(credentialsProvider::refresh);
         }
     }
 
-    @Override
-    public void close() {
-        releaseCachedClients();
+    /**
+     * Customizes {@link com.amazonaws.auth.WebIdentityTokenCredentialsProvider}
+     *
+     * <ul>
+     * <li>Reads the the location of the web identity token not from AWS_WEB_IDENTITY_TOKEN_FILE, but from a symlink
+     * in the plugin directory, so we don't need to create a hardcoded read file permission for the plugin.</li>
+     * <li>Supports customization of the STS endpoint via a system property, so we can test it against a test fixture.</li>
+     * <li>Supports gracefully shutting down the provider and the STS client.</li>
+     * </ul>
+     */
+    static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentialsProvider {
+
+        private STSAssumeRoleWithWebIdentitySessionCredentialsProvider credentialsProvider;
+        private AWSSecurityTokenService stsClient;
+
+        CustomWebIdentityTokenCredentialsProvider(Environment environment) {
+            // Check whether the original environment variable exists. If it doesn't,
+            // the system doesn't support AWS web identity tokens
+            if (System.getenv(AWS_WEB_IDENTITY_ENV_VAR) == null) {
+                return;
+            }
+            // Make sure that a readable symlink to the token file exists in the plugin config directory
+            Path webIdentityTokenFileSymlink = environment.configFile().resolve("repository-s3/aws-web-identity-token-file");
+            if (Files.exists(webIdentityTokenFileSymlink) == false) {
+                throw new IllegalStateException(
+                    "A Web Identity Token symlink in the config directory doesn't exist: " + webIdentityTokenFileSymlink
+                );
+            }
+            if (Files.isReadable(webIdentityTokenFileSymlink) == false) {
+                throw new IllegalStateException(
+                    "Unable to read a Web Identity Token symlink in the config directory: " + webIdentityTokenFileSymlink
+                );
+            }
+            String roleArn = System.getenv(AWS_ROLE_ARN_ENV_VAR);
+            String roleSessionName = System.getenv(AWS_ROLE_SESSION_NAME_ENV_VAR);
+            if (roleArn == null || roleSessionName == null) {
+                return;
+            }
+            AWSSecurityTokenServiceClientBuilder stsClientBuilder = AWSSecurityTokenServiceClient.builder();
```

Comment:
Just to be sure to understand - this client is shared among all repositories or should we have one instance per repository? Should we also be able to configure its S3 client configuration (timeouts, retries, max connections etc)?

Reply:
The former: one instance per repository. We can expose its client configuration, but I'm not sure it's necessary now unless somebody asks for them. It's only used for communicating with the AWS STS and it uses reasonable default `ClientConfiguration` with timeouts and retries.

Reply:
> one instance per repository

Does that mean that all S3 repositories will use STS as long as the symlink exists?